ReactNativeClient/lib/services/searchengine/queryBuilder.ts

```diff
@@ -0,0 +1,505 @@
+		});
+		const matchQuery = (relation === 'OR') ? termsToMatch.join(' OR ') : termsToMatch.join(' ');
+		params.push(matchQuery);
+	}
+
+	const excludedTextTerms = allTerms.filter(x => x.name === 'text' && x.negated);
+	const excludedTitleTerms = allTerms.filter(x => x.name === 'title' && x.negated);
+	const excludedBodyTerms = allTerms.filter(x => x.name === 'body' && x.negated);
+
+	if ((excludedTextTerms.length > 0) && relation === 'AND') {
+		conditions.push(`
+		AND ROWID NOT IN (
+		SELECT ROWID
+		FROM notes_fts
+		WHERE notes_fts MATCH ?
+		)`);
+		params.push(excludedTextTerms.map(x => x.value).join(' OR '));
+	}
+
+	if ((excludedTextTerms.length > 0) && relation === 'OR') {
+		excludedTextTerms.map(textTerm => {
+			conditions.push(`
+			OR ROWID IN (
+			SELECT *
+			FROM (
+				SELECT ROWID
+				FROM notes_fts
+				EXCEPT
+				SELECT ROWID FROM notes_fts
+				WHERE notes_fts MATCH ?
+			))`);
+			params.push(textTerm.value);
+		});
+	}
+
+	if (excludedTitleTerms.length > 0) {
+		addExcludeTextConditions(excludedTitleTerms, conditions, params, relation);
+	}
+
+	if (excludedBodyTerms.length > 0) {
+		addExcludeTextConditions(excludedBodyTerms, conditions, params, relation);
+	}
+
+};
+
+const getDefaultRelation = (filters: Term[]): 'OR' | 'AND' => {
+	const anyTerm = filters.find(term => term.name === 'any');
+	if (anyTerm) { return (anyTerm.value === '1') ? 'OR' : 'AND'; }
+	return 'AND';
+};
+
+const getConnective = (filters: Term[], relation: 'OR' | 'AND'): string => {
+	const notebookTerm = filters.find(x => x.name === 'notebook');
+	return (!notebookTerm && (relation === 'OR')) ? 'ROWID=-1' : '1'; // ROWID=-1 acts as 0 (something always false)
+};
+
+export default function queryBuilder(filters: Term[]) {
+	const queryParts: string[] = [];
+	const params: string[] = [];
+	const withs: string[] = [];
+
+	// console.log("testing beep beep boop boop")
+	// console.log(filters);
+
+	const relation: 'AND' | 'OR' = getDefaultRelation(filters);
+
+	// console.log(`relations = ${relation}`);
+	// console.log(filters);
+
+	queryParts.push(`
+	SELECT
+	notes_fts.id,
+	notes_fts.title,
+	offsets(notes_fts) AS offsets,
+	notes_fts.user_created_time,
+	notes_fts.user_updated_time,
+	notes_fts.is_todo,
+	notes_fts.todo_completed,
+	notes_fts.parent_id
+	FROM notes_fts
+	WHERE ${getConnective(filters, relation)}`);
+
+	notebookFilter(filters, withs, queryParts, params);
+
+	tagFilter(filters, withs, queryParts, params, relation);
+
+	resourceFilter(filters, withs, queryParts, params, relation);
+
+	textFilter(filters, queryParts, params, relation);
+
+	typeFilter(filters, queryParts, relation);
+
+	completedFilter(filters, queryParts, relation);
+
+	dateFilter(filters, queryParts, params, relation);
+
+	locationFilter(filters, queryParts, params, relation);
+
+	let query;
+	if (withs.length > 0) {
+		query = ['WITH RECURSIVE' , withs.join(',') ,queryParts.join(' ')].join(' ');
+	} else {
+		query = queryParts.join(' ');
+	}
```

Comment:
Thanks for cleaning up the code Naveen, this is much more readable now. However, I still see a lot of almost duplicated code. For example the content of dateFilter and typeFilter, completedFilter and locationFilter is almost identical. The resourceFilter and tagFilter also repeats several time the same blocks of code, with small differences between both.

I feel the code needs to be abstracted more, to make it easier to see what's common between the filters, and what are the differences. Like it took me a while to see why there was different conditional blocks in tagFilter or resourceFilter because the code is so similar. All this makes it difficult to know what the code does, and will make it difficult to maintain it.

Do you have any suggestions on how you could simplify the code further?